src/BotwinExtensions.cs

```diff
@@ -21,6 +21,13 @@ public static class BotwinExtensions
         /// <returns>A reference to this instance after the operation has completed.</returns>
         public static IApplicationBuilder UseBotwin(this IApplicationBuilder builder, BotwinOptions options = null)
         {
+            var diagnostics = builder.ApplicationServices.GetService<BotwinDiagnostics>();
+            if (diagnostics != null)
+            {
+                var logger = builder.ApplicationServices.GetService<ILoggerFactory>().CreateLogger(typeof(BotwinDiagnostics));
+                diagnostics.LogDiscoveredModules(logger);
```

Comment:
Rather than adding types to lists would it be more performant to just resolve modules, validators etc and log them here? I don't really like that as it seems a perf hit.

I know the api currently is a bit yucky but its a bit more performant

Reply:
As mentioned above, this would have been an alternative implementation.

Collecting the types in a list will only require some additional memory, although I would not expect `Type` references to be heavy.

On the other hand, resolving all types once at the beginning will actively create those objects, and all their dependencies. Since there are no restrictions on what users could do and what things they could depend on, I would expect this to have a bigger impact. And there’s a possibility of side effects too. Not to mention that resolving objects just to throw them away feel really bad 🙈 

As far as performance goes, this should not really make any difference, as this only ever runs once when the application starts up. So this would only affect the startup time (although I wouldn’t expect this to be noticeable).